### PR TITLE
Update PHP tests for compatibility with PHP 8 and WP 5.9+ and WP < 5.2

### PR DESCRIPTION
**Pull request by aaemnnosttv**
## Summary

Addresses issue:

- #4729 

The "wp-phpunit" portion of WordPress core changed with WP 5.9 to allow for compatibility with newer versions of PHPUnit which are incompatible with PHP below PHP 7 due to the addition of return types on fixture methods (e.g. `setUp`, `tearDown` etc). Omitting these return types raises errors due to incompatible definitions which forced WP core to adopt the [PHPUnit Polyfills](https://github.com/Yoast/PHPUnit-Polyfills) library. This allows for authoring test cases that are forward-compatible with newer versions of PHPUnit.

This does not come for free however and requires us to rewrite our fixture methods to match the snake-case (cross-version compatible) names used by WordPress core's base test case.

This is further complicated by the changes made to the WP core base test case only being back ported to WP 5.2 which leaves the new methods uncalled in tests based on wp-phpunit for WP below 5.2.

The solution here was to introduce an additional polyfill layer between Site Kit's base test case and the `WP_UnitTestCase` by using a new `WP_UnitTestCase_Polyfill` class which works very similar to way the PHPUnit polyfill test case works only a bit simpler. If the `WP_UnitTestCase` class does not have the newer methods, we load a polyfill version of the class which is basically a copy of the same class used in WP 5.2 of the library. Otherwise, we load an empty "pass-through" version of the class; both then extend the `WP_UnitTestCase` class. This allows us to restore the same functionality for our test case while maintaining support for wp-phpunit with our current minimum supported version of WP (4.7).

The added polyfill has been implemented as a local path-type Composer package, to make it easier to extract/replace later.

The rest of the changes here are to fix deprecation warnings with v8 and v9 of PHPUnit which otherwise fail, even though no test failures occur. Using PHPUnit 8 seemed to be required as PHPUnit 7 was failing with parse errors ("unexpected token match"). The lift for compatibility from PHPUnit 8 to 9 was trivial so this was included as well. 

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: .github/workflows/php-tests-wp-nightly-php-7-4.yml [hunk 1]
```diff
@@ -75,4 +75,4 @@ jobs:
       - name: Set up PHP test data
         run: tests/bin/install-wp-tests.sh ${MYSQL_DATABASE} ${MYSQL_USER} ${MYSQL_PASSWORD} ${DB_HOST}:${DB_PORT} ${WP_VERSION}
       - name: Run Unit Tests
-        run: composer test
+        run: composer test -- --verbose
```
Comment (felixarntz) [inline .github/workflows/php-tests-wp-nightly-php-7-4.yml:78]: Not opposed, but curious why you changed this throughout.

File: tests/phpunit/bootstrap.php [hunk 2]
```diff
@@ -9,6 +9,8 @@
  */
 
 define( 'TESTS_PLUGIN_DIR', dirname( dirname( __DIR__ ) ) );
+// Ensure Composer autoloader is available.
+require_once TESTS_PLUGIN_DIR . '/vendor/autoload.php';
```
Comment (felixarntz) [inline tests/phpunit/bootstrap.php:13]: Moving this here seems a bit unrelated, was there a specific problem that led you to doing that? In any case, I agree having that here outside of the condition below makes more sense.

**Review by felixarntz — APPROVED** [review]
@aaemnnosttv Amazing work!

While it obviously has to be a bit clunky, this implementation is as clean as it can be and works reliably. I left two minor comments, but nothing blocking.

File: .github/workflows/php-tests-wp-nightly-php-7-4.yml (hunk 1 above)
Comment (aaemnnosttv) [reply .github/workflows/php-tests-wp-nightly-php-7-4.yml:78]: Mostly because it adds information about the runtime to the run:

```
Runtime:       PHP 8.0.14
Configuration: /home/runner/work/site-kit-wp/site-kit-wp/phpunit.xml.dist
```

It also adds things like the names of tests that were incomplete or have been skipped.

If we really wanted to make our output more verbose (like Jest) we could use `--testdox`, but `--verbose` doesn't really add that much. I mainly wanted to expose the specific version of PHP used.

File: tests/phpunit/bootstrap.php (hunk 2 above)
Comment (aaemnnosttv) [reply tests/phpunit/bootstrap.php:13]: It's now needed in all cases (not only when using wp-phpunit locally) since we use it to load the PHPUnit Polyfills library, but also the added `packages/wp-phpunit-polyfill` package which is loaded by Composer.

Comment (aaemnnosttv) [reply tests/phpunit/bootstrap.php]: Turns out this actually isn't needed at all because PHPUnit loads it automatically 😄

**Comment by aaemnnosttv** [comment]
Thanks @felixarntz if there's anything else, let's address it in a follow-up PR/issue 👍
